### ISSUE TITLE
Implement generalized Pipek-Mezey

### ISSUE THREAD

**Issue by susilehtola** [pyscf/pyscf#563]
I was hoping to get [generalized Pipek-Mezey localization](https://pubs.acs.org/doi/10.1021/ct401016x) in PySCF. Basically this boils down to replacing the Mulliken charges with something that has a basis set limit, and you can use whatever you like as different choices turn out to give very closely the same orbitals as we showed in the original paper and its followup on [Pipek-Mezey Wannier functions](https://pubs.acs.org/doi/10.1021/acs.jctc.6b00809).

The simplest way to do this is to define the atomic population matrix P^A_{ij} in terms of atomic overlap matrices S^A_{ij} = \int w_A(r) \phi_i(r) \phi_j(r) d^3 r, which can be built with the density functional routines. The simplest choice for the weights is the Becke quadrature, meaning that you can get the grid to build S^A_{ij} for an atom by modifying gen_partition to build only the grid for a single atom https://github.com/pyscf/pyscf/blob/8015c2dbdf242200368e8b29b16a784bd8ec43a8/pyscf/dft/gen_grid.py#L335

In short, the algorithm in metacode
```
loop over atoms A:
     form Becke grid for A only
     evaluate basis functions phi
     form atomic overlap S^A = phi * grid weight * phi.T
     store population matrix P^A = C.T * S^A * C
```
This is an extremely simple algorithm. However, given that there's a bunch of parameters you need to pass on to the DFT routines (radial scheme, pruning, etc) I'm not sure how to organize the code within PySCF. I'd be equally happy if someone who already knows the code would put this in.

**Comment by sunqm** [comment]
Is it possible to adapt your proposed population matrix into the PM implementation https://github.com/pyscf/pyscf/blob/master/pyscf/lo/pipek.py#L33 ?

**Comment by susilehtola** [comment]
@sunqm I think so; the code would be something like
```
for A in range(natoms):
    S_A = numint.atomic_overlap(A)
    proj[A] = numpy.dot(mo_coeff.conj().T, numpy.dot(S_A, mo_coeff))
```
The question is just how to get the atomic overlap matrices by quadrature from the DFT code. `atomic_overlap` should only evaluate the grid centered on atom `A`.

**Comment by sunqm** [comment]
Does `atomic_overlap` have a concrete boundary in real space for each atom or something like the becke grids centered on the atom with some grids sitting at the region of other atoms?

**Comment by susilehtola** [comment]
`atomic_overlap` is only evaluating the integral for atom `A`, defined by the Becke grid weighting scheme.

**Comment by susilehtola** [comment]
Oh great, you were fast. Thanks!
